### PR TITLE
Strict mode

### LINKED ISSUE #236: `const` and strict mode
According to the [ES6 draft](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-let-and-const-declarations), `const` variables are now block-scoped (scoped in the same way as `let` variables:

> `let` and `const` declarations define variables that are scoped to the [running execution context](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-execution-contexts)’s [LexicalEnvironment](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-execution-contexts).

[In contrast](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-variable-statement) to `var`, which is scoped to a script or function:

> A `var` statement declares variables that are scoped to the [running execution context](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-execution-contexts)’s [VariableEnvironment](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-execution-contexts).

When `const` was just a non-standard Mozilla extension, and while in ES6 non-strict mode, it is scoped the same as `var`.

It might be a good idea to check if the current scope is running in strict mode so that `const` can be added to the correct scope. I expect the variables to be scoped like this:

<p align="center"> </p>

But right now in `js2-mode` they are scoped like this (in strict mode the `c` is still blue):

<p align="center"> </p>

I've taken a look at the source and haven't found any indication that ECMAScript strict mode can be detected. There is a `js2-mode-show-strict-warnings` flag which claims to show "Ecma strict-mode warnings", but these warnings include trailing commas, missing semicolons, inconsistent return values, variable redeclarations, ironically *none* of which are actually observed by `"use strict";`.

Therefore, I'd like to request *real* strict mode support.

(For the implementation, note that `"use strict"` must appear as the first statement in a function:)

```js
(function () { 'use strict'; {const a = 1;} console.log(a); }());
// Uncaught ReferenceError: a is not defined

(function () { void 0; 'use strict'; {const a = 1;} console.log(a); }());
// 1
```

### PR DESCRIPTION
**Pull request by jacksonrayhamilton**
Resolves #236, adding strict mode support to the parser.

- Scripts and functions now have a slot for `in-strict-mode`.
- Functions inherit strict mode.
- `const` variables are now scoped correctly according to context.

I was feeling generous, so I also implemented the rest of Rhino's strict mode warnings:

- `eval` and `arguments` are illegal identifiers.
- Duplicate names in function parameters is an error.
- `with` statement is an error.
- Octal syntax is an error.
- Duplicate keys in an object literal is an error.

I am pretty happy with this, except for the retroactive function name and parameter strict mode error checking, which Rhino didn't even seem to bother with. I'd gladly accept suggestions on simplifying that. I considered looking ahead at the function body for a `use strict` directive and then jumping back to parse the name and parameters, but I wasn't quite sure how to "pass over" function parameters and backed off.

### CODE REVIEW

**Comment by dgutov** [comment]
Hi Jackson,

This is fantastic, I'll try to review it soon.

> I am pretty happy with this, except for the retroactive function name and parameter strict mode error checking, which Rhino didn't even seem to bother with. I'd gladly accept suggestions on simplifying that.

Have you looked at what SpiderMonkey does? IIRC, it saves the token stream state before parsing a function, and then rewinds if it needs to be reparsed in strict mode. Also see the use of `js2-ts-seek` in `js2-parse-assign-expr`.

File: js2-mode.el [hunk 1]
```diff
@@ -7923,14 +7957,18 @@ leaving a statement, an expression, or a function definition."
 Scanner should be initialized."
   (let ((pos js2-ts-cursor)
         (end js2-ts-cursor)  ; in case file is empty
-        root n tt)
+        root n tt
+        (in-directive-prologue t)
+        (saved-strict-mode js2-in-use-strict-directive)
```
Comment (dgutov) [inline js2-mode.el]: I'd rather bind `js2-in-use-strict-directive` here so itself, then it'll save us the `saved-strict-mode` variable, and the value will be restored upon exiting the `let` form.

File: js2-mode.el [hunk 2]
```diff
@@ -8166,20 +8301,29 @@ arrow function), NAME is js2-name-node."
...
+           js2-loop-and-switch-set
+           (error-checking-arguments (js2-parse-function-params
+                                      function-type fn-node pos)))
       (when (eq function-type 'FUNCTION_ARROW)
         (js2-must-match js2-ARROW "msg.bad.arrow.args"))
       (if (and (>= js2-language-version 180)
                (/= (js2-peek-token) js2-LC))
           (js2-parse-function-closure-body fn-node)
         (js2-parse-function-body fn-node))
+      (when (js2-function-node-in-strict-mode fn-node)
```
Comment (dgutov) [inline js2-mode.el]: Couldn't you perform this logic while `js2-in-use-strict-directive` is still assigned? Then there will be no need for the `in-strict-mode` slot.

**Comment by jacksonrayhamilton** [comment]
I took the SpiderMonkey approach and rewound the parser. Much less egregious now.

Also removed the strict mode slots. I suspected they weren't needed (and they weren't after let-binding), just thought they might be useful for consumers. If someone wants them we can add them back.

**Comment by dgutov** [comment]
Thank you.

> If someone wants them we can add them back.

Sounds good.

Now, please open https://raw.githubusercontent.com/jashkenas/underscore/master/underscore.js and add `"use strict";` at the top of the file. After that, I'm getting this error:

```
Debugger entered--Lisp error: (wrong-type-argument js2-name-node [cl-struct-js2-prop-get-node 33 1512 13 nil nil 1516 [cl-struct-js2-keyword-node 43 0 4 nil #1] [cl-struct-js2-name-node 39 5 8 nil #1 "_wrapped" nil]])
  signal(wrong-type-argument (js2-name-node [cl-struct-js2-prop-get-node 33 1512 13 nil nil 1516 [cl-struct-js2-keyword-node 43 0 4 nil #1] [cl-struct-js2-name-node 39 5 8 nil #1 "_wrapped" nil]]))
  (or (progn nil (and (vectorp name-node) (>= (length name-node) 8) (memq (aref name-node 0) cl-struct-js2-name-node-tags) t)) (signal (quote wrong-type-argument) (list (quote js2-name-node) name-node)))
  (progn nil (or (progn nil (and (vectorp name-node) (>= (length name-node) 8) (memq (aref name-node 0) cl-struct-js2-name-node-tags) t)) (signal (quote wrong-type-argument) (list (quote js2-name-node) name-node))) (aref name-node 6))
  (let ((param-name (progn nil (or (progn nil (and (vectorp name-node) (>= ... 8) (memq ... cl-struct-js2-name-node-tags) t)) (signal (quote wrong-type-argument) (list (quote js2-name-node) name-node))) (aref name-node 6)))) (if (member param-name js2-illegal-strict-identifiers) (progn (js2-report-error "msg.bad.id.strict" param-name (js2-node-abs-pos name-node) (progn nil (or (progn nil (and ... ... ... t)) (signal (quote wrong-type-argument) (list ... name-node))) (aref name-node 3))))))
  (progn (let ((param-name (progn nil (or (progn nil (and ... ... ... t)) (signal (quote wrong-type-argument) (list ... name-node))) (aref name-node 6)))) (if (member param-name js2-illegal-strict-identifiers) (progn (js2-report-error "msg.bad.id.strict" param-name (js2-node-abs-pos name-node) (progn nil (or (progn nil ...) (signal ... ...)) (aref name-node 3)))))))
  (if js2-in-use-strict-directive (progn (let ((param-name (progn nil (or (progn nil ...) (signal ... ...)) (aref name-node 6)))) (if (member param-name js2-illegal-strict-identifiers) (progn (js2-report-error "msg.bad.id.strict" param-name (js2-node-abs-pos name-node) (progn nil (or ... ...) (aref name-node 3))))))))
  js2-check-strict-identifier([cl-struct-js2-prop-get-node 33 1512 13 nil nil 1516 [cl-struct-js2-keyword-node 43 0 4 nil #0] [cl-struct-js2-name-node 39 5 8 nil #0 "_wrapped" nil]])
  (cond ((and (<= js2-first-assign tt) (<= tt js2-last-assign)) (setq op-pos (- (js2-current-token-beg) pos) left pn) (js2-check-strict-identifier left) (setq right (js2-parse-assign-expr) pn (make-js2-assign-node :type tt :pos pos :len (- (js2-node-end right) pos) :op-pos op-pos :left left :right right)) (if js2-parse-ide-mode (progn (js2-highlight-assign-targets pn left right) (js2-record-imenu-functions right left))) (js2-node-add-children pn left right)) ((and (= tt js2-ARROW) (>= js2-language-version 200)) (js2-ts-seek ts-state) (setq js2-recorded-identifiers recorded-identifiers js2-parsed-errors parsed-errors) (setq pn (js2-parse-function (quote FUNCTION_ARROW) (js2-current-token-beg) nil))) (t (js2-unget-token)))
  (if (= tt js2-YIELD) (js2-parse-return-or-yield tt t) (setq ts-state (progn nil (vector (quote cl-struct-js2-ts-state) js2-ts-lineno js2-ts-cursor (copy-sequence js2-ti-tokens) js2-ti-tokens-cursor js2-ti-lookahead)) recorded-identifiers js2-recorded-identifiers parsed-errors js2-parsed-errors) (setq pn (js2-parse-cond-expr) tt (js2-get-token)) (cond ((and (<= js2-first-assign tt) (<= tt js2-last-assign)) (setq op-pos (- (js2-current-token-beg) pos) left pn) (js2-check-strict-identifier left) (setq right (js2-parse-assign-expr) pn (make-js2-assign-node :type tt :pos pos :len (- (js2-node-end right) pos) :op-pos op-pos :left left :right right)) (if js2-parse-ide-mode (progn (js2-highlight-assign-targets pn left right) (js2-record-imenu-functions right left))) (js2-node-add-children pn left right)) ((and (= tt js2-ARROW) (>= js2-language-version 200)) (js2-ts-seek ts-state) (setq js2-recorded-identifiers recorded-identifiers js2-parsed-errors parsed-errors) (setq pn (js2-parse-function (quote FUNCTION_ARROW) (js2-current-token-beg) nil))) (t (js2-unget-token))) pn)
  (let ((tt (js2-get-token)) (pos (js2-current-token-beg)) pn left right op-pos ts-state recorded-identifiers parsed-errors) (if (= tt js2-YIELD) (js2-parse-return-or-yield tt t) (setq ts-state (progn nil (vector (quote cl-struct-js2-ts-state) js2-ts-lineno js2-ts-cursor (copy-sequence js2-ti-tokens) js2-ti-tokens-cursor js2-ti-lookahead)) recorded-identifiers js2-recorded-identifiers parsed-errors js2-parsed-errors) (setq pn (js2-parse-cond-expr) tt (js2-get-token)) (cond ((and (<= js2-first-assign tt) (<= tt js2-last-assign)) (setq op-pos (- (js2-current-token-beg) pos) left pn) (js2-check-strict-identifier left) (setq right (js2-parse-assign-expr) pn (make-js2-assign-node :type tt :pos pos :len (- (js2-node-end right) pos) :op-pos op-pos :left left :right right)) (if js2-parse-ide-mode (progn (js2-highlight-assign-targets pn left right) (js2-record-imenu-functions right left))) (js2-node-add-children pn left right)) ((and (= tt js2-ARROW) (>= js2-language-version 200)) (js2-ts-seek ts-state) (setq js2-recorded-identifiers recorded-identifiers js2-parsed-errors parsed-errors) (setq pn (js2-parse-function (quote FUNCTION_ARROW) (js2-current-token-beg) nil))) (t (js2-unget-token))) pn))
  js2-parse-assign-expr()
  (let* ((pn (js2-parse-assign-expr)) (pos (progn nil (or (progn nil (and (vectorp pn) (>= ... 6) (memq ... cl-struct-js2-node-tags) t)) (signal (quote wrong-type-argument) (list (quote js2-node) pn))) (aref pn 2))) left right op-pos) (while (and (not oneshot) (js2-match-token js2-COMMA)) (setq op-pos (- (js2-current-token-beg) pos)) (if (= (js2-peek-token) js2-YIELD) (js2-report-error "msg.yield.parenthesized")) (setq right (js2-parse-assign-expr) left pn pn (progn "Represents infix expressions.\nIncludes assignment ops like `|=', and the comma operator.\nThe type field inherited from `js2-node' holds the operator." nil (vector (quote cl-struct-js2-infix-node) js2-COMMA pos (- js2-ts-cursor pos) nil nil op-pos left right))) (js2-node-add-children pn left right)) pn)
  js2-parse-expr()
  (progn (js2-unget-token) (js2-parse-expr))
  (js2-wrap-with-expr-stmt (js2-current-token-beg) (progn (js2-unget-token) (js2-parse-expr)) t)
  js2-parse-expr-stmt()
...
```

**Comment by jacksonrayhamilton** [comment]
Should be fixed now.

**Comment by dgutov** [comment]
:+1:

**Pull request merged**